<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Изучение-общей-информации" data-toc-modified-id="Изучение-общей-информации-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Изучение общей информации</a></span><ul class="toc-item"><li><span><a href="#Испорт-необходимых-библиотек" data-toc-modified-id="Испорт-необходимых-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Испорт необходимых библиотек</a></span></li><li><span><a href="#Открытие-файлов" data-toc-modified-id="Открытие-файлов-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Открытие файлов</a></span></li><li><span><a href="#Вывод-первых-десяти-строк-таблиц" data-toc-modified-id="Вывод-первых-десяти-строк-таблиц-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод первых десяти строк таблиц</a></span></li><li><span><a href="#Вывод-основной-информации-о-датафреймах" data-toc-modified-id="Вывод-основной-информации-о-датафреймах-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Вывод основной информации о датафреймах</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Разбиение-данных-на-целевой-признак-и-признаки" data-toc-modified-id="Разбиение-данных-на-целевой-признак-и-признаки-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Разбиение данных на целевой признак и признаки</a></span></li><li><span><a href="#Разбиение-данных-на-выборки" data-toc-modified-id="Разбиение-данных-на-выборки-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Разбиение данных на выборки</a></span></li><li><span><a href="#Масштабирование-признаков" data-toc-modified-id="Масштабирование-признаков-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Масштабирование признаков</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обучение-и-проверка-модели" data-toc-modified-id="Обучение-и-проверка-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение и проверка модели</a></span><ul class="toc-item"><li><span><a href="#Обучение-модели-и-предсказания" data-toc-modified-id="Обучение-модели-и-предсказания-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Обучение модели и предсказания</a></span></li><li><span><a href="#Среднее-значение-предсказаний-и-RMSE" data-toc-modified-id="Среднее-значение-предсказаний-и-RMSE-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Среднее значение предсказаний и RMSE</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Подготовка-к-расчёту-прибыли" data-toc-modified-id="Подготовка-к-расчёту-прибыли-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Подготовка к расчёту прибыли</a></span><ul class="toc-item"><li><span><a href="#Создание-констант" data-toc-modified-id="Создание-констант-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Создание констант</a></span></li><li><span><a href="#Расчет-достаточного-объема-сырья" data-toc-modified-id="Расчет-достаточного-объема-сырья-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Расчет достаточного объема сырья</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Расчёт-прибыли-и-рисков" data-toc-modified-id="Расчёт-прибыли-и-рисков-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Расчёт прибыли и рисков</a></span><ul class="toc-item"><li><span><a href="#Функция-для-расчета-прибыли" data-toc-modified-id="Функция-для-расчета-прибыли-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Функция для расчета прибыли</a></span></li><li><span><a href="#Процедура-Bootstrap" data-toc-modified-id="Процедура-Bootstrap-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Процедура Bootstrap</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. 

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самыми высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

**Условия задачи:**
- Для обучения модели подходит только линейная регрессия (остальные — недостаточно предсказуемые).
- При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки.
- Бюджет на разработку скважин в регионе — 10 млрд рублей.
- При нынешних ценах один баррель сырья приносит 450 рублей дохода. Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
- После оценки рисков нужно оставить лишь те регионы, в которых вероятность убытков меньше 2.5%. Среди них выбирают регион с наибольшей средней прибылью.

**Цель исследования** — построить модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализировать возможную прибыль и риски техникой *Bootstrap.* 

**Ход исследования:**
1. Изучение файлов
2. Подготовка данных
3. Обучение и проверка моделей
4. Подготовка к расчету прибыли
5. Расчет прибыли и рисков
6. Вывод

## Изучение общей информации

### Испорт необходимых библиотек

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split 

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

### Открытие файлов

In [2]:
try:
    data_1 = pd.read_csv('/datasets/geo_data_0.csv', engine='python')
    data_2 = pd.read_csv('/datasets/geo_data_1.csv', engine='python')
    data_3 = pd.read_csv('/datasets/geo_data_2.csv', engine='python')
except:
    data_1 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_0.csv', engine='python')
    data_2 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_1.csv', engine='python')
    data_3 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_2.csv', engine='python')

### Вывод первых десяти строк таблиц

In [3]:
data_1.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [4]:
data_2.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [5]:
data_3.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

### Вывод основной информации о датафреймах

In [6]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [7]:
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [8]:
data_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


### Вывод

Согласно документации к данным:
- `id` — уникальный идентификатор скважины;
- `f0, f1, f2` — три признака точек (неважно, что они означают, но сами признаки значимы);
- `product` — объём запасов в скважине (тыс. баррелей).

Целевым признаком будет являться `product`, а признаками — `f1`, `f2`, `f3`. Необходимо разбить данные каждого датафрейма на обучающую и валидационную выборки в отношении 75:25 и промасштабировать признаки.

## Подготовка данных

### Разбиение данных на целевой признак и признаки

In [9]:
def split(data):
    features = data.drop(['id', 'product'], axis=1)
    target = data['product']
    
    return features, target

features_1, target_1 = split(data_1)
features_2, target_2 = split(data_2)
features_3, target_3 = split(data_3)

### Разбиение данных на выборки

In [10]:
def train_valid(features, target):
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=12345)
    return features_train, features_valid, target_train, target_valid

features_train_1, features_valid_1, target_train_1, target_valid_1 = train_valid(features_1, target_1)
features_train_2, features_valid_2, target_train_2, target_valid_2 = train_valid(features_2, target_2)
features_train_3, features_valid_3, target_train_3, target_valid_3 = train_valid(features_3, target_3)

In [11]:
print(features_train_1.shape, features_valid_1.shape)
print(features_train_2.shape, features_valid_2.shape)
print(features_train_3.shape, features_valid_3.shape)

(75000, 3) (25000, 3)
(75000, 3) (25000, 3)
(75000, 3) (25000, 3)


### Масштабирование признаков

In [12]:
numeric = ['f1', 'f2', 'f2']

pd.options.mode.chained_assignment = None

def scale(features_train, features_valid):
    scaler = StandardScaler()
    scaler.fit(features_train[numeric])
    
    features_train[numeric] = scaler.transform(features_train[numeric])
    features_valid[numeric] = scaler.transform(features_valid[numeric])
    
    return features_train, features_valid

features_train_1, features_valid_1 = scale(features_train_1, features_valid_1)
features_train_2, features_valid_2 = scale(features_train_2, features_valid_2)
features_train_3, features_valid_3 = scale(features_train_3, features_valid_3)

### Вывод

На этапе подготовки данные были разделены на тренировочную и валидационную выборки в отношении 75:25: в тренировочной выборке 75000 объектов, а в валидационной 25000 объектов. Также понадобилось промасштабировать численные признаки.

## Обучение и проверка модели

### Обучение модели и предсказания

In [13]:
def model_fit(features_train, target_train, features_valid):
    model = LinearRegression()
    model.fit(features_train, target_train)
    
    predictions = model.predict(features_valid)
    
    return predictions

predictions_1 = model_fit(features_train_1, target_train_1, features_valid_1)
predictions_2 = model_fit(features_train_2, target_train_2, features_valid_2)
predictions_3 = model_fit(features_train_3, target_train_3, features_valid_3)

### Среднее значение предсказаний и RMSE

In [14]:
def mean_rmse(predictions, target):
    mean = predictions.mean()
    rmse = mean_squared_error(target, predictions, squared=False)
    
    return mean, rmse
    
print('Для первого датафрейма среднее значение предсказаний равно %f, а RMSE равно %f'%mean_rmse(predictions_1, target_valid_1))
print('Для второго датафрейма среднее значение предсказаний равно %f, а RMSE равно %f'%mean_rmse(predictions_2, target_valid_2))
print('Для третьего датафрейма среднее значение предсказаний равно %f, а RMSE равно %f'%mean_rmse(predictions_3, target_valid_3))

Для первого датафрейма среднее значение предсказаний равно 92.592568, а RMSE равно 37.579422
Для второго датафрейма среднее значение предсказаний равно 68.728547, а RMSE равно 0.893099
Для третьего датафрейма среднее значение предсказаний равно 94.965046, а RMSE равно 40.029709


### Вывод

Самое низкое значение RMSE у второго датафрейма, у третьего — самое высокое. Так как RMSE измеряется в тех же величинах, что и исходные значения, модель для второго датафрейма работает корректнее всего.

## Подготовка к расчёту прибыли

### Создание констант

In [15]:
RESEARCH_AMOUNT = 500 #количество точек для исследования
WORKING_AMOUNT = 200 #количество точек для разработки
BUDGET = 10000000000 #бюджет, выделенный на разработку
UNIT_INCOME = 450000 #доход от единицы продукта

### Расчет достаточного объема сырья

In [16]:
SUFFICIENT = BUDGET / (UNIT_INCOME * WORKING_AMOUNT)
SUFFICIENT

111.11111111111111

### Вывод

По расчетам достаточный объем сырья для безубыточной разработки новой скважины равен 111 тысячам, что больше, чем средние значения всех датафреймов.

## Расчёт прибыли и рисков

### Функция для расчета прибыли

In [17]:
def profit(target, predictions):
    predictions_sorted = predictions.sort_values(ascending=False)
    selected = target[predictions_sorted.index][:WORKING_AMOUNT]
    
    return UNIT_INCOME * selected.sum() - BUDGET

### Процедура Bootstrap

In [18]:
state = np.random.RandomState(12345)

def bootstrap(target, predictions):
    values = []

    for i in range(1000):
        target_subsample = target.reset_index(drop=True).sample(n=RESEARCH_AMOUNT, replace=True, random_state=state)
        predictions_subsample = pd.Series(predictions)[target_subsample.index] 
        income = profit(target_subsample, predictions_subsample)
        values.append(income)
        
    values = pd.Series(values)
    mean_income = values.mean()
        
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)
    
    risk = (values < 0).mean()
        
    return mean_income, lower, upper, risk

In [19]:
mean_income, lower, upper, loss = bootstrap(target_valid_1, predictions_1)

print('Для первого датафрейма:')
print('Средняя прибыль:', mean_income)
print('95%-й доверительный интервал:', lower, '-', upper) 
print('Риск убытков:', loss)

Для первого датафрейма:
Средняя прибыль: 425938526.91059244
95%-й доверительный интервал: -102090094.83793654 - 947976353.3583689
Риск убытков: 0.06


In [20]:
mean_income, lower, upper, loss = bootstrap(target_valid_2, predictions_2)

print('Для второго датафрейма:')
print('Средняя прибыль:', mean_income)
print('95%-й доверительный интервал:', lower, '-', upper) 
print('Риск убытков:', loss)

Для второго датафрейма:
Средняя прибыль: 518259493.6973249
95%-й доверительный интервал: 128123231.43308444 - 953612982.0669085
Риск убытков: 0.003


In [21]:
mean_income, lower, upper, loss = bootstrap(target_valid_3, predictions_3)

print('Для третьего датафрейма:')
print('Средняя прибыль:', mean_income)
print('95%-й доверительный интервал:', lower, '-', upper) 
print('Риск убытков:', loss)

Для третьего датафрейма:
Средняя прибыль: 420194005.3440501
95%-й доверительный интервал: -115852609.16001143 - 989629939.8445739
Риск убытков: 0.062


### Вывод

Самым прибыльным регионом является второй: он имеет самую высокую среднюю прибыль, самые низкие убытки и положительный доверительный интервал.

##  Вывод

Данные датафреймов были подготовлены к работе с моделями: разбиты на целевые признаки и признаки, затем разбиты на обучающие и валидационные выборки в отношении 75:25, а численные признаки были промасштабированы.

Далее была обучена модель линейной регрессии и сделаны предсказания, по котором была высчитана RMSE. По данным показателям наименьшую ошибку выдала вторая модель.

По расчетам из четвертого пункта было выявлено, что минимальный безубыточный объем равен 111 тысячам баррелей, что выше средних значений по объемам в датафреймах.

Через процедуру bootstrap были рассчитаны средняя прибыль, 95%-доверительный интервал и риск убытков для каждого региона. Самым прибыльным является второй, так как имеет самую высокую среднюю прибыль, самые низкие убытки и положительный доверительный интервал.